peer/peer.go

```diff
@@ -0,0 +1,2043 @@
+	notFoundMsgListeners    map[string]func(*Peer, *wire.MsgNotFound)
+	getDataMsgListeners     map[string]func(*Peer, *wire.MsgGetData)
+	getBlocksMsgListeners   map[string]func(*Peer, *wire.MsgGetBlocks)
+	getHeadersMsgListeners  map[string]func(*Peer, *wire.MsgGetHeaders)
+	filterAddMsgListeners   map[string]func(*Peer, *wire.MsgFilterAdd)
+	filterClearMsgListeners map[string]func(*Peer, *wire.MsgFilterClear)
+	filterLoadMsgListeners  map[string]func(*Peer, *wire.MsgFilterLoad)
+	versionMsgListeners     map[string]func(*Peer, *wire.MsgVersion)
+	rejectMsgListeners      map[string]func(*Peer, *wire.MsgReject)
+}
+
+// String returns the peer's address and directionality as a human-readable
+// string.
+func (p *Peer) String() string {
+	return fmt.Sprintf("%s (%s)", p.addr, directionString(p.inbound))
+}
+
+// isKnownInventory returns whether or not the peer is known to have the passed
+// inventory.  It is safe for concurrent access.
+func (p *Peer) isKnownInventory(invVect *wire.InvVect) bool {
+	if p.knownInventory.Exists(invVect) {
+		return true
+	}
+	return false
+}
+
+// UpdateLastBlockHeight updates the last known block for the peer. It is safe
+// for concurrent access.
+func (p *Peer) UpdateLastBlockHeight(newHeight int32) {
+	p.statsMtx.Lock()
+	defer p.statsMtx.Unlock()
+
+	log.Tracef("Updating last block height of peer %v from %v to %v",
+		p.addr, p.lastBlock, newHeight)
+	p.lastBlock = int32(newHeight)
+}
+
+// UpdateLastAnnouncedBlock updates meta-data about the last block sha this
+// peer is known to have announced. It is safe for concurrent access.
+func (p *Peer) UpdateLastAnnouncedBlock(blkSha *wire.ShaHash) {
+	p.statsMtx.Lock()
+	defer p.statsMtx.Unlock()
+
+	log.Tracef("Updating last blk for peer %v, %v", p.addr, blkSha)
+	p.lastAnnouncedBlock = blkSha
+}
+
+// AddKnownInventory adds the passed inventory to the cache of known inventory
+// for the peer.  It is safe for concurrent access.
+func (p *Peer) AddKnownInventory(invVect *wire.InvVect) {
+	p.knownInventory.Add(invVect)
+}
+
+// AddKnownAddress adds the passed address to the cache of known addresses
+// for the peer.  It is safe for concurrent access.
+func (p *Peer) AddKnownAddress(addr string) {
+	p.knownAddrMutex.Lock()
+	defer p.knownAddrMutex.Unlock()
+
+	p.knownAddresses[addr] = struct{}{}
+}
+
+// ID returns the peer id.
+func (p *Peer) ID() int32 {
+	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
+
+	return p.id
+}
+
+// NA returns the peer network address.
+func (p *Peer) NA() *wire.NetAddress {
+	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
+
+	return p.na
+}
+
+// Addr returns the peer address.
+func (p *Peer) Addr() string {
+	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
+
+	return p.addr
+}
+
+// Filter returns the peer bloom filter.
+func (p *Peer) Filter() *bloom.Filter {
+	p.flagsMtx.Lock()
+	defer p.flagsMtx.Unlock()
+
+	return p.filter
+}
+
+// Inbound returns whether the peer is inbound.
+func (p *Peer) Inbound() bool {
+	p.flagsMtx.Lock()
```

Comment:
Same deal here.  The `inbound` field is assigned at creation time and never modified.  No need for mutex.